src/main/java/com/ebay/myriad/scheduler/NMHeartBeatHandler.java

```diff
@@ -0,0 +1,155 @@
+      OfferLifecycleManager offerLifecycleMgr,
+      NodeStore nodeStore) {
+
+    if (registry != null) {
+      registry.register(this);
+    }
+
+    this.yarnScheduler = yarnScheduler;
+    this.myriadDriver = myriadDriver;
+    this.yarnNodeCapacityMgr = yarnNodeCapacityMgr;
+    this.offerLifecycleMgr = offerLifecycleMgr;
+    this.nodeStore = nodeStore;
+  }
+
+  @Override
+  public void beforeRMNodeEventHandled(RMNodeEvent event, RMContext context) {
+    switch (event.getType()) {
+      case STARTED: {
+        // TODO (Santosh) We can't zero out resources here in all cases. For e.g.
+        // this event might be fired when an existing node is rejoining the
+        // cluster (NM restart) as well. Sometimes this event can have a list of
+        // container statuses, which, capacity/fifo schedulers seem to handle
+        // (perhaps due to work preserving NM restart).
+        RMNode rmNode = context.getRMNodes().get(event.getNodeId());
+        yarnNodeCapacityMgr.removeCapacity(rmNode);
+      }
+      break;
+
+      case EXPIRE: {
+        nodeStore.remove(event.getNodeId().getHost());
+      }
+      break;
+
+      case STATUS_UPDATE: {
+        handleStatusUpdate(event, context);
+      }
+      break;
+
+      default: {
+        LOGGER.debug("Unhandled event: {}", event.getClass().getName());
+      }
+      break;
+    }
+  }
+
+  private void handleStatusUpdate(RMNodeEvent event, RMContext context) {
+    if (!(event instanceof RMNodeStatusEvent)) {
+      LOGGER.error("{} not an instance of {}", event.getClass().getName(),
+          RMNodeStatusEvent.class.getName());
+      return;
+    }
+
+    RMNodeStatusEvent statusEvent = (RMNodeStatusEvent) event;
+    RMNode rmNode = context.getRMNodes().get(event.getNodeId());
+    SchedulerNode schedulerNode = yarnScheduler.getSchedulerNode(rmNode.getNodeID());
+    String hostName = rmNode.getNodeID().getHost();
+
+    int rmSchNodeNumContainers = schedulerNode.getNumContainers();
+    List<ContainerStatus> nmContainers = statusEvent.getContainers();
+
+    if (nmContainers.size() != rmSchNodeNumContainers) {
+      LOGGER.warn("Node: {}, Num Containers known by RM scheduler: {}, "
+          + "Num containers NM is reporting: {}",
+          hostName, rmSchNodeNumContainers, nmContainers.size());
+    }
+
+    // Send task update to Mesos
+    for (ContainerStatus status : nmContainers) {
+      ContainerId containerId = status.getContainerId();
+      Protos.SlaveID slaveId = nodeStore.getNode(hostName).getSlaveId();
+      if (status.getState() == ContainerState.COMPLETE) {
+        LOGGER.debug("Task complete: {}", containerId);
+        yarnNodeCapacityMgr.removeCapacity(rmNode, containerId);
+        requestExecutorToSendTaskStatusUpdate(
+            slaveId,
+            containerId, Protos.TaskState.TASK_FINISHED);
+      } else { // state == NEW | RUNNING
+        // TODO (Kannan) Can we not let the executor send this update by itself?
```

Comment:
Executor currently doesn't know the status of the YARN container. So, the framework sends this message to the executor to send back a status update to mesos.